examples/singleCheckpointTest/main.cc

```diff
@@ -0,0 +1,204 @@
+    }
+
+    std::cout << "DUMMY (" << mpiHelper.rank()
+              << "): Initialize preCICE and set mesh\n";
+    couplingParticipant.setMesh(meshName, vertices);
+
+    // Create index mapping between DuMuX's index numbering and preCICE's numbering
+    std::cout << "DUMMY (" << mpiHelper.rank() << "): Create index mapping\n";
+    couplingParticipant.createIndexMapping(dumuxVertexIDs);
+
+    couplingParticipant.announceQuantity(meshName, dataToWrite);
+    couplingParticipant.announceQuantity(meshName, dataToRead);
+
+    if (couplingParticipant.requiresToWriteInitialData()) {
+        std::cout << "DUMMY (" << mpiHelper.rank()
+                  << "): Writing initial data\n";
+        couplingParticipant.writeQuantityVector(meshName, dataToWrite,
+                                                writeScalarData);
+        couplingParticipant.writeQuantityToOtherSolver(meshName, dataToWrite);
+    }
+    std::cout << "DUMMY (" << mpiHelper.rank() << "): Exchange initial\n";
+    couplingParticipant.initialize();
+    double preciceDt = 0;
+
+    // Create instances of the file-scope mock types for checkpointing.
+    GridVariables gridVars;
+    TimeLoop timeLoop;
+
+    // Register writeScalarData as the solver state for checkpointing.
+    couplingParticipant.initializeCheckpoint(writeScalarData, gridVars,
+                                             timeLoop);
+
+    // Check exchanged initial data
+    if (solverName == "SolverOne") {
+        std::cout << "SolverOne: Reading initial data\n";
+        couplingParticipant.readQuantityFromOtherSolver(meshName, dataToRead,
+                                                        preciceDt);
+    }
+
+    int iter = 0;
+    dataToKeep = writeScalarData;
+    double timeToKeep = timeLoop.time();
+    long timeStepIndexToKeep = timeLoop.timeStepIndex();
+
+    while (couplingParticipant.isCouplingOngoing()) {
+        if (solverName == "SolverOne") {
+            couplingParticipant.writeCheckpointIfRequired();
+
+            ++iter;
+
+            std::iota(writeScalarData.begin(), writeScalarData.end(), iter);
+
+            preciceDt = couplingParticipant.getMaxTimeStepSize();
+            couplingParticipant.advance(preciceDt);
+            timeLoop.advanceTime(preciceDt);
+            if (!couplingParticipant.readCheckpointIfRequired()) {
+                timeToKeep = timeLoop.time();
+                timeStepIndexToKeep = timeLoop.timeStepIndex();
+            }
+            if (writeScalarData != dataToKeep) {
+                throw std::runtime_error(
+                    "SolverOne: Checkpointing failed, data not restored "
+                    "correctly");
+            }
+            if ((timeStepIndexToKeep != timeLoop.timeStepIndex()) ||
+                (timeToKeep != timeLoop.time())) {
+                throw std::runtime_error(
+                    "SolverOne: Checkpointing failed, time step not "
+                    "restored correctly");
+            }
```

Comment:
```suggestion
            if (couplingParticipant.readCheckpointIfRequired()) {
                if (writeScalarData != dataToKeep) {
                    throw std::runtime_error(
                        "SolverOne: Checkpointing failed, data not restored "
                        "correctly");
                }
                if ((timeStepIndexToKeep != timeLoop.timeStepIndex()) ||
                    (timeToKeep != timeLoop.time()) ||
                    (timeStepSizeToKeep != timeLoop.timeStepSize())) {
                    throw std::runtime_error(
                        "SolverOne: Checkpointing failed, time step not "
                        "restored correctly");
                }
            }
```
Storing comparison data should happen when checkpointing (see above). Check reset only if actually resetting. Also check for time-step size.